### ISSUE TITLE
Ktor client cannot catch offline exceptions (Kotlin Multiplatform)

### ISSUE THREAD

**Issue by LiewJunTung** [ktorio/ktor#1165]
### Ktor Version
1.2.0

### Ktor Engine Used(client or server and name)
IOS Client

### JVM Version, Operating System and Relevant Context
iOS - Airplane mode

### Feedback
Tried to catch offline mode exception in case the internet connection is dropped. The errors showed up in the logs but cannot be caught in Multiplatform common project. Haven't tried on Android yet.

```
2019-06-03 16:02:17.432609+0800 XApp[1558:1117673] Task <449DDD76-DA3D-42B4-9CA4-ECEA0B218850>.<1> load failed with error Error Domain=NSURLErrorDomain Code=-1009 "The Internet connection appears to be offline." UserInfo={_kCFStreamErrorCodeKey=50, NSUnderlyingError=0x282e59d40 {Error Domain=kCFErrorDomainCFNetwork Code=-1009 "(null)" UserInfo={_kCFStreamErrorCodeKey=50, _kCFStreamErrorDomainKey=1}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <449DDD76-DA3D-42B4-9CA4-ECEA0B218850>.<1>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <449DDD76-DA3D-42B4-9CA4-ECEA0B218850>.<1>"
), NSLocalizedDescription=The Internet connection appears to be offline., NSErrorFailingURLStringKey=https://localhost/?page=1&ordering=modified&search=, NSErrorFailingURLKey=https://localhost/?page=1&ordering=modified&search=, _kCFStreamErrorDomainKey=1} [-1009]
```

**Comment by tonilopezmr** [comment]
I have the same problem, @LiewJunTung how did you catch this exception?

**Comment by LiewJunTung** [comment]
@tonilopezmr for now I'm not handling it. And will wait for the fix by @e5l. :)

**Comment by flaquir4** [comment]
I had the same problem. But I've solved it by catching a Throwable instead of an Exception in the try/catch block.

Hope it helps.

**Comment by tonilopezmr** [comment]
I think I was catching with Throwable as well :think: I'll try another time

**Comment by svendroid** [comment]
I am experiencing the same problem, catching Throwable instead of Exception did not work for me.

**Comment by nuno-vieira** [comment]
This only happens in release mode. Also true for you guys?

**Comment by sergiocasero** [comment]
Throwable works nice for me, but the problem I have is that I show the "Default error".  Is there any way to know what error is?, I'm doing something like this:

```kotlin
private suspend fun <R> execute(f: suspend () -> R): Either<Error, R> =
        try {
            Either.Right(f())
        } catch (e: Throwable) {
            print(e.toString())
            Either.Left(
                when (e) {
                    is ClientRequestException -> when (e.response.status) {
                        HttpStatusCode.Unauthorized -> Error.InvalidCredentials
                        HttpStatusCode.NotFound -> Error.NotFound
                        HttpStatusCode.BadRequest -> Error.NoInternet
                        else -> Error.Default
                    }
                    else -> Error.Default
                }
            )
        }```

**Comment by cy6erGn0m** [comment]
Should be fixed in 1.3.0. Could you please verify?

**Comment by sergiocasero** [comment]
Hello @cyrillhalter , which should be the exception type? ClientRequestException?

**Comment by vanniktech** [comment]
I also just ran into this using the simulator and using ktor 1.3.0:

```
2020-02-05 15:46:31.381391+0100 FAMILO_ALPHA[30817:536791] Connection 8: received failure notification
2020-02-05 15:46:31.381514+0100 FAMILO_ALPHA[30817:536791] Connection 8: failed to connect 1:50, reason -1
2020-02-05 15:46:31.381594+0100 FAMILO_ALPHA[30817:536791] Connection 8: encountered error(1:50)
2020-02-05 15:46:31.383628+0100 FAMILO_ALPHA[30817:536791] Task <4FFC1C58-A2B1-4BAD-B250-90A3E0364287>.<1> HTTP load failed, 0/0 bytes (error code: -1009 [1:50])
2020-02-05 15:46:31.385023+0100 FAMILO_ALPHA[30817:536677] Task <4FFC1C58-A2B1-4BAD-B250-90A3E0364287>.<1> finished with error [-1009] Error Domain=NSURLErrorDomain Code=-1009 "The Internet connection appears to be offline." UserInfo={_kCFStreamErrorCodeKey=50, NSUnderlyingError=0x600003284f00 {Error Domain=kCFErrorDomainCFNetwork Code=-1009 "(null)" UserInfo={_kCFStreamErrorCodeKey=50, _kCFStreamErrorDomainKey=1}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <4FFC1C58-A2B1-4BAD-B250-90A3E0364287>.<1>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <4FFC1C58-A2B1-4BAD-B250-90A3E0364287>.<1>"
), NSLocalizedDescription=The Internet connection appears to be offline., NSErrorFailingURLStringKey=https://dev-api.familonet-api.net/api/v4/authenticate/anonymous/, NSErrorFailingURLKey=https://dev-api.familonet-api.net/api/v4/authenticate/anonymous/, _kCFStreamErrorDomainKey=1}
[Http] REQUEST https://dev-api.familonet-api.net/api/v4/authenticate/anonymous/ failed with exception: io.ktor.client.engine.ios.IosHttpRequestException: Exception in http request: Error Domain=NSURLErrorDomain Code=-1009 "The Internet connection appears to be offline." UserInfo={_kCFStreamErrorCodeKey=50, NSUnderlyingError=0x600003284f00 {Error Domain=kCFErrorDomainCFNetwork Code=-1009 "(null)" UserInfo={_kCFStreamErrorCodeKey=50, _kCFStreamErrorDomainKey=1}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <4FFC1C58-A2B1-4BAD-B250-90A3E0364287>.<1>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <4FFC1C58-A2B1-4BAD-B250-90A3E0364287>.<1>"
), NSLocalizedDescription=The Internet connection appears to be offline., NSErrorFailingURLStringKey=https://dev-api.familonet-api.net/api/v4/authenticate/anonymous/, NSErrorFailingURLKey=https://dev-api.familonet-api.net/api/v4/authenticate/anonymous/, _kCFStreamErrorDomainKey=1} 
io.ktor.client.engine.ios.IosHttpRequestException: Exception in http request: Error Domain=NSURLErrorDomain Code=-1009 "The Internet connection appears to be offline." UserInfo={_kCFStreamErrorCodeKey=50, NSUnderlyingError=0x600003284f00 {Error Domain=kCFErrorDomainCFNetwork Code=-1009 "(null)" UserInfo={_kCFStreamErrorCodeKey=50, _kCFStreamErrorDomainKey=1}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <4FFC1C58-A2B1-4BAD-B250-90A3E0364287>.<1>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <4FFC1C58-A2B1-4BAD-B250-90A3E0364287>.<1>"
), NSLocalizedDescription=The Internet connection appears to be offline., NSErrorFailingURLStringKey=https://dev-api.familonet-api.net/api/v4/authenticate/anonymous/, NSErrorFailingURLKey=https://dev-api.familonet-api.net/api/v4/authenticate/anonymous/, _kCFStreamErrorDomainKey=1}
        at 0   FAMILO_ALPHA                        0x000000010eb2bf67 kfun:kotlin.Throwable.<init>(kotlin.String?)kotlin.Throwable + 87
        at 1   FAMILO_ALPHA                        0x000000010ede4e6f kfun:io.ktor.client.engine.ios.IosHttpRequestException.<init>(platform.Foundation.NSError?)io.ktor.client.engine.ios.IosHttpRequestException + 431
        at 2   FAMILO_ALPHA                        0x000000010ede2380 kfun:io.ktor.client.engine.ios.IosClientEngine.<no name provided>_6.URLSession_10#internal + 2832
        at 3   FAMILO_ALPHA                        0x000000010ede40bb _knbridge60 + 379
        at 4   CFNetwork                           0x00007fff22e9c0dc _CFNetworkHTTPConnectionCacheSetLimit + 206684
        at 5   Foundation                          0x00007fff2567de0f __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 7
        at 6   Foundation                          0x00007fff2567dd07 -[NSBlockOperation main] + 98
        at 7   Foundation                          0x00007fff25680c94 __NSOPERATION_IS_INVOKING_MAIN__ + 17
        at 8   Foundation                          0x00007fff2567cf07 -[NSOperation start] + 731
        at 9   Foundation                          0x00007fff256815ee __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 17
        at 10  Foundation                          0x00007fff256810f8 __NSOQSchedule_f + 182
        at 11  libdispatch.dylib                   0x00000001115ee407 _dispatch_block_async_invoke2 + 83
        at 12  libdispatch.dylib                   0x00000001115dfd6c _dispatch_client_callout + 8
        at 13  libdispatch.dylib                   0x00000001115ede24 _dispatch_main_queue_callback_4CF + 1500
        at 14  CoreFoundation                      0x00007fff23b0ce49 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
        at 15  CoreFoundation                      0x00007fff23b07aa9 __CFRunLoopRun + 2329
        at 16  CoreFoundation                      0x00007fff23b06e66 CFRunLoopRunSpecific + 438
        at 17  GraphicsServices                    0x00007fff38346bb0 GSEventRunModal + 65
        at 18  UIKitCore                           0x00007fff47578dd0 UIApplicationMain + 1621
        at 19  FAMILO_ALPHA                        0x000000010e8196cf main + 143 (/Users/niklas/dev/familonet/familonet-ios/Sources/FamilonetAPI/Swift/main.swift:4:1)
        at 20  libdyld.dylib                       0x00007fff516ecd29 start + 1
        at 21  ???                                 0x0000000000000001 0x0 + 1
```

**Comment by icarumbas** [comment]
Another workaround is to use CoroutineExceptionHandler

**Comment by dcoletto** [comment]
Same issue here with android. I get a "java.net.UnknownHostException" when there's no internet connection (eg: airplane mode) which is of course only available on jvm, so I can't catch it on the common side. I think it should be possible to handle the exception on the platform specific side, but I'd like to know if that's just a "temporary workaround" or it is the desired behaviour

**Comment by Alex009** [comment]
i think reason of error is this line - https://github.com/ktorio/ktor/blob/63c508918543774b857977ad033d602e04491dc4/ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosUtils.kt#L40
`IosHttpRequestException` not `Exception` but just `Throwable`. it's reason why we not catch this `Exception` when use `catch(e: Exception)` and why `catch(e: Throwable)` will help.
i think this case should be Exception, not just Throwable.

**Comment by Ahmy000** [comment]
@e5l  Yes, it's more reasonable that IosHttpRequestException is an Exception.

**Comment by joselufo** [comment]
@e5l when it's going to be released the fix?

**Comment by nbransby** [comment]
> `[IosHttpRequestException parent to improve usability](/ktorio/ktor/commit/6732b4a09efb23bac390098b7b76da2a4d91bbac) ` …

And if you are using CIO on Android you get an `java.nio.channels.UnresolvedAddressException` which again you can't catch in common code, is there a solution for this part of the issue yet?

**Comment by SandeepDhami2242** [comment]
> Same issue here with android. I get a "java.net.UnknownHostException" when there's no internet connection (eg: airplane mode) which is of course only available on jvm, so I can't catch it on the common side. I think it should be possible to handle the exception on the platform specific side, but I'd like to know if that's just a "temporary workaround" or it is the desired behaviour

Any updates here?

**Comment by e5l** [comment]
@SandeepDhami2242, could you log a separate issue for that?

**Comment by bestMechanic** [comment]
Any update on this? I am facing an issue with the request as submitFormWithBinaryData. throwable is also not able to capture this.

**Comment by bestMechanic** [comment]
> This only happens in release mode. Also true for you guys?

Did you find any solution for this, looks like this is happening in release mode.

**Comment by nuno-vieira** [comment]
> 

@bestMechanic I'm not working for the same company, nor I'm using KMM anymore. But I remember that we stopped using Ktor for iOS since we encountered so many issues and unexpected crashes. We implemented our own wrapper around URLSession for the iOS KMM actual implementation. So we used the actual/expected thing for the networking code. For Android, we used Ktor, for iOS, we used our own URLSession wrapper.
 
Since I left, I know that the company completely dropped KMM as well since it did not work well for our use case.
